Close #96 Add a 'Choosing the Right Branch for Your Quickstart Project' section… (#120)
az-digital/az-quickstart-scaffolding: README.md

# Arizona Quickstart Scaffolding

## Notes
If you are planning on pushing this site to Pantheon, you should use the
[Pantheon upstream repository](https://github.com/az-digital/az-quickstart-pantheon) as your scaffolding repo,
which has separate instructions.

![Security workflow](https://github.com/az-digital/az-quickstart-scaffolding/workflows/Security%20workflow/badge.svg)

- Designed to build a Quickstart website/project codebase
- Includes [Quickstart](https://github.com/az-digital/az_quickstart) as dependency
- Installs Drupal in a `web` subdirectory
- Utilizes [composer-installers](https://github.com/composer/installers) and [Drupal Composer Scaffold](https://github.com/drupal/core-composer-scaffold) plugins

> - Can be used as project template for composer [create-project command](https://getcomposer.org/doc/03-cli.md#create-project):

Reviewer: ```suggestion
- Can be used as a project template for the composer [create-project command](https://getcomposer.org/doc/03-cli.md#create-project)
```

**Notes**
- If developing a feature for `az-digital/az_quickstart`, use the provided build tools (DDev or Lando) instead of running `composer create-project` directly.
- These branches are not as stable as the main or release branches, so use
  them for development, experimentation, or testing rather than in production
  environments.
- The use of feature or issue branches in this repository is tightly aligned
  with developments in the az-digital/az_quickstart project, and they should be
  used in conjunction with the corresponding Quickstart branches.

### Using the `composer create-project` command:

With Composer's `create-project` command, you can quickly scaffold a new project
using the Arizona Quickstart Scaffolding. Each `az-quickstart-scaffolding`
branch coincides with an `az_quickstart` branch, ensuring compatibility and
alignment between the scaffolding and the core project.

Here are some examples of how you might use this command with different branch
specifications:

If using any of the following commands, adjust `my_project_name` to your desired
project directory name.

> - **Using the Main Branch (`main`)**:

Author: ```suggestion
- **Using the Main Branch (`main`) with dev dependencies**:
```

   ```bash

>    composer create-project az-digital/az-quickstart-scaffolding:dev-main my_project_name --no-interaction --no-dev

Author: ```suggestion
   composer create-project az-digital/az-quickstart-scaffolding:dev-main my_project_name --no-interaction
```

> 2. **Using the Latest Unreleased Version Alias (`2.8.x`):**

Author: ```suggestion
- **Using the Latest Unreleased Version Alias (`2.8.x`):**
```

   ```bash
   composer create-project az-digital/az-quickstart-scaffolding:2.8.x-dev my_project_name --no-interaction --no-dev
   ```

> 3. **Using a Specific Release Branch (e.g., `2.7.x`):**

Author: ```suggestion
- **Using a Specific Release Branch (e.g., `2.7.x`):**
```

  ```bash
  composer create-project az-digital/az-quickstart-scaffolding:2.7.x-dev my_project_name --no-interaction --no-dev
  ```

> 4. **Using an Issue or Feature Branch (e.g., `issue-45`)**

Author: ```suggestion
- **Using an Issue or Feature Branch (e.g., `issue-45`)**
```

- Installs Drupal in a `web` subdirectory
- Utilizes [composer-installers](https://github.com/composer/installers) and [Drupal Composer Scaffold](https://github.com/drupal/core-composer-scaffold) plugins
- Can be used as project template for composer [create-project command](https://getcomposer.org/doc/03-cli.md#create-project):

## Notes
If you are planning on pushing this site to Pantheon, you should use the
[Pantheon upstream repository](https://github.com/az-digital/az-quickstart-pantheon) as your scaffolding repo,
which has separate instructions.

## Choosing the Right Branch for Your Quickstart Project
In the Arizona Quickstart Scaffolding project, different branches serve various
purposes. Here's a guide to understanding which branch you should choose:

### Main Branch (`dev-main`)

- **Purpose**: The main branch contains all the latest features and changes. It

>   represents the cutting-edge development but may be less stable than specific

Author: ```suggestion
  represents the cutting-edge development and may be less stable than specific
```

  release branches. You can use it with `dev-main` or an alias for the latest
  unreleased version (e.g., `2.8.x`).
- **When to Use**: Choose this branch if you want access to the very latest
  features, and you are willing to work with potentially less-tested code. Ideal

>   for development, experimentation, or testing new functionalities.

Author: ```suggestion
  for development, experimentation, or testing unreleased functionality.
```

> - **Purpose**: The main branch contains all the latest features and changes. It
>   represents the cutting-edge development and may be less stable than specific
>   release branches. You can use it with `dev-main` or an alias for the latest
>   unreleased version (e.g., `2.8.x`).
> - **When to Use**: Choose this branch if you want access to the very latest
>   features, and you are willing to work with potentially less-tested code. Ideal
>   for development, experimentation, or testing new functionalities.

Author: Alternate 
```suggestion
- **Purpose**: The main branch downloads the latest supported version of `az-digital/az_quickstart` and contains unreleased updates to the scaffolding repository. While it provides access to the latest stable features of Quickstart, it also includes cutting-edge changes to the scaffolding that may be less stable.
- **When to Use**: Choose this branch if you need a stable version of Quickstart for development or testing, and you want to work with the latest updates and improvements to the scaffolding itself.
```

Author: Should probably also mention that it only downloads the latest stable version at the time, and until the next minor release comes out.

> 

Author: ```suggestion
```

  achieving a robust and reliable implementation.
- **Version Constraints in Supported Release Branches**: Each supported release
  branch in this repository specifies a version constraint for az_quickstart in composer.json.
  This constraint ensures that the release branch will work with compatible
  versions of Quickstart that correspond to the compatible minor release branch,
  providing flexibility while maintaining alignment with the intended versions.

### Feature or Issue Branches

Feature or issue branches in this repository are typically created to make
updates to the scaffolding to work with specific branches of
`az-digital/az_quickstart`. When creating such a branch, it is customary to pin
this repository's composer.json reference for `az-digital/az_quickstart` to a
specific issue/feature branch that the changes are intended to work with. This
ensures that the branch's changes align precisely with a particular state of the
Quickstart project.

These branches are useful for isolated development, testing of changes, or adding new features to the scaffolding that correspond to developments in the Quickstart project.

**Notes**

> - These branches are not as stable as the main or release branches, so use

Author: ```suggestion
- If developing a feature for `az-digital/az_quickstart`, use the provided build tools (DDev or Lando) instead of running `composer create-project` directly.
- These branches are not as stable as the main or release branches, so use
```

  them for development, experimentation, or testing rather than in production
  environments.
- The use of feature or issue branches in this repository is tightly aligned
  with developments in the az-digital/az_quickstart project, and they should be
  used in conjunction with the corresponding Quickstart branches.

### Using the `composer create-project` command:

With Composer's `create-project` command, you can quickly scaffold a new project
using the Arizona Quickstart Scaffolding. Each `az-quickstart-scaffolding`
branch coincides with an `az_quickstart` branch, ensuring compatibility and
alignment between the scaffolding and the core project.

Here are some examples of how you might use this command with different branch
specifications:

If using any of the following commands, adjust `my_project_name` to your desired
project directory name.

1. **Using the Main Branch (`main`)**:
   ```bash

>    composer create-project az-digital/az-quickstart-scaffolding:dev-main my_project_name --no-interaction --no-dev

Author: Should we recommend --no-dev here?

>    composer create-project az-digital/az-quickstart-scaffolding:dev-main my_project_name --no-interaction --no-dev

Author: ```suggestion
   composer create-project az-digital/az-quickstart-scaffolding:dev-main my_project_name --no-interaction
```

Author: This suggestion only makes sense if we change the scaffolding repo to use az-digital/az_quickstart: ~2.8.x

2. **Using the Latest Unreleased Version Alias (`2.8.x`):**

   ```bash
   composer create-project az-digital/az-quickstart-scaffolding:2.8.x-dev my_project_name --no-interaction --no-dev
   ```

3. **Using a Specific Release Branch (e.g., `2.7.x`):**
  ```bash
  composer create-project az-digital/az-quickstart-scaffolding:2.7.x-dev my_project_name --no-interaction --no-dev
  ```

4. **Using an Issue or Feature Branch (e.g., `issue-45`)**

  ```bash

>   composer create-project az-digital/az-quickstart-scaffolding:dev-issue/45 my_project_name --no-interaction --no-dev

Author: ```suggestion
  composer create-project az-digital/az-quickstart-scaffolding:dev-issue/45 my_project_name --no-interaction
```

>   ```

Author: Should we recommend --no-dev here?

- Installs Drupal in a `web` subdirectory
- Utilizes [composer-installers](https://github.com/composer/installers) and [Drupal Composer Scaffold](https://github.com/drupal/core-composer-scaffold) plugins
- Can be used as project template for composer [create-project command](https://getcomposer.org/doc/03-cli.md#create-project):

## Notes
If you are planning on pushing this site to Pantheon, you should use the
[Pantheon upstream repository](https://github.com/az-digital/az-quickstart-pantheon) as your scaffolding repo,

> which has separate instructions.

Author: Should this remain at the top?

- **Version Constraints in Supported Release Branches**: Each supported release
  branch in this repository specifies a version constraint for az_quickstart in composer.json.
  This constraint ensures that the release branch will work with compatible
  versions of Quickstart that correspond to the compatible minor release branch,
  providing flexibility while maintaining alignment with the intended versions.
### Feature or Issue Branches

Feature or issue branches in this repository are typically created to make
updates to the scaffolding to work with specific branches of
`az-digital/az_quickstart`. When creating such a branch, it is customary to pin
this repository's composer.json reference for `az-digital/az_quickstart` to a
specific issue/feature branch that the changes are intended to work with. This
ensures that the branch's changes align precisely with a particular state of the
Quickstart project.

These branches are useful for isolated development, testing of changes, or adding new features to the scaffolding that correspond to developments in the Quickstart project.

**Notes**
- Feature or issue branches may also be referred to as issue branches, depending

>   on the context or the specific workflow of the project.

Reviewer 2: Since you always refer to "feature or issue branches", this comment seems unnecessary.

- These branches may not be as stable as the main or release branches, so use
  them for development, experimentation, or testing rather than in production
  environments.
- The use of feature or issue branches in this repository is tightly aligned
  with developments in the az-digital/az_quickstart project, and they should be
  used in conjunction with the corresponding Quickstart branches.

### Using the `composer create-project` command:

With Composer's `create-project` command, you can quickly scaffold a new project
using the Arizona Quickstart Scaffolding. Each `az-quickstart-scaffolding`
branch coincides with an `az_quickstart` branch, ensuring compatibility and
alignment between the scaffolding and the core project.

Here are some examples of how you might use this command with different branch
specifications:

If using any of the following commands, adjust `my_project_name` to your desired
project directory name.

1. **Using the Main Branch (`main`)**:
   ```bash
   composer create-project az-digital/az-quickstart-scaffolding:dev-main my_project_name --no-interaction --no-dev

> 2. **Using the Latest Unreleased Version Alias (`2.8.x`):**

Reviewer 2: Are we going to be updating this file every time we create a new release?

- Installs Drupal in a `web` subdirectory
- Utilizes [composer-installers](https://github.com/composer/installers) and [Drupal Composer Scaffold](https://github.com/drupal/core-composer-scaffold) plugins
- Can be used as project template for composer [create-project command](https://getcomposer.org/doc/03-cli.md#create-project):

## Notes
If you are planning on pushing this site to Pantheon, you should use the
[Pantheon upstream repository](https://github.com/az-digital/az-quickstart-pantheon) as your scaffolding repo,
which has separate instructions.

## Choosing the Right Branch for Your Quickstart Project
In the Arizona Quickstart Scaffolding project, different branches serve various purposes. Here's a guide to understanding which branch you should choose:

### Main Branch (`dev-main`, `2.8.x`)

> - **Purpose**: The main branch contains all the latest features and changes. It represents the cutting-edge development but may be less stable than specific release branches. You can use it with `dev-main` or an alias for the latest unreleased version (e.g., `2.8.x`).

Reviewer 3: This is not consistent with the current contents of `composer.json` in the main branch, correct?  (Since it curently requires `az-digital/az_quickstart:~2.7`)

> - **Purpose**: The main branch contains all the latest features and changes. It represents the cutting-edge development but may be less stable than specific release branches. You can use it with `dev-main` or an alias for the latest unreleased version (e.g., `2.8.x`).

Author: True, I got a little confused, thinking I was writing about quickstart.

> - **Purpose**: The main branch contains all the latest features and changes. It represents the cutting-edge development but may be less stable than specific release branches. You can use it with `dev-main` or an alias for the latest unreleased version (e.g., `2.8.x`).

Author: I guess that is what you were talking about in the morning meeting when you were saying that we should change that to the dev-main alias.

I think with the explanation in these docs, we could switch to using the dev-main alias.

> - **When to Use**: Choose this branch if you want access to the very latest features, and you are willing to work with potentially less-tested code. Ideal for development, experimentation, or testing new functionalities.

Author: This part only makes sense if we change the scaffolding repo to use az-digital/az_quickstart: ~2.8.x
